listener.go

```diff
@@ -70,7 +81,8 @@ func (l *listener) handleConn(conn net.Conn) {
 	if l.KeepAliveInterval > 0 {
 		smuxConfig.KeepAliveInterval = l.KeepAliveInterval
 	}
-	session, err := smux.Server(conn, smuxConfig)
+	teeConn := newTeeConn(conn)
```

Comment:
If I understand this correctly, we're going to read everything sent to this conn into an in-memory buffer? That seems not optimal. Once we get past accepting the first stream, we don't really need this mechanism anymore, so we don't need to keep buffering into memory.

I ran into a similar problem with MITM processing and solved it here - https://github.com/getlantern/proxy/blob/master/proxy_http.go#L229

Basically, we keep track of what we've read until the point where we make the go/no go decision. On no go, we return a [preconn](https://godoc.org/github.com/getlantern/preconn) that inserts the already read data at the head of the underlying connection.

Reply:
Oh that's a great point!

Okay, I'll fix it and thank you for the pointers, that's helpful.

Reply:
Okay, how's this?

https://github.com/getlantern/cmux/commit/68fc2124d0f0accbbbdca7cfc5974357d09fc2a0

A little different than your suggestion, but it should work and I think it's straight-forward enough.

Reply:
Made a minor improvement here: https://github.com/getlantern/cmux/commit/2404e13e4c040f30549b021eb098a3c51820900b

Now there will be no more references to the buffer after `stopTee` is called, so it should get garbage collected.